<a href="https://colab.research.google.com/github/JulioHenri/Analise-Restaurante/blob/master/tec_amostragem.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importando bibliotecas

In [1]:
import pandas as pd
import numpy as np

# Carregando e lendo o banco de dados

In [2]:
dados = pd.read_csv('HR_Engagement_Sat_Sales_UpdatedV4.0.csv') #carregando arquivo csv
dados.head() #lendo os 5 primeiros registros do dataset

ID  ... Sensor_Proximity(1-highest/10-lowest)
0   1  ...                                     9
1   2  ...                                     8
2   3  ...                                     8
3   4  ...                                     6
4   5  ...                                    10

[5 rows x 62 columns]

In [3]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 62 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   ID                                     14999 non-null  int64  
 1   Name                                   14997 non-null  object 
 2   Department                             14999 non-null  object 
 3   GEO                                    14999 non-null  object 
 4   Role                                   14999 non-null  object 
 5   Rising_Star                            21 non-null     float64
 6   Will_Relocate                          14999 non-null  int64  
 7   Critical                               480 non-null    float64
 8   Trending Perf                          14999 non-null  int64  
 9   Talent_Level                           14999 non-null  int64  
 10  Validated_Talent_Level                 14999 non-null  int64  
 11  Pe

Podemos notar acima que há 14999 registros e 62 colunas. Também há várias atributos com valores faltantes que serão analisados posteriormente.

# Exercícios

Exercício 1.1

a. unidade de pesquisa: Se o empregado continua ou não na empresa

b. população: Funcionários de uma empresa

c. instrumento de coleta de dados: Sistema de cadastro da empresa

d. unidade respondente: Empresas na área de Recursos Humanos

e. possível sistema de referência: 

f. unidade amostral mais provável: Apenas funcionários com salário "baixo"

g. unidades amostrais alternativas: Funcionários do departamento de Finanças e  os que não foram promovidos nos últimos 5 anos

Exercício 1.5 

a. Árvores em uma floresta: Nesse caso, iria fazer a amostra aleatória sistemática pois ela é de rápida e fácil execução e dessa maneira poderíamos selecionar as diversas árvores e separa-las por características específicas e depois de ter feito isso, e depois escolher um número igual de árvores para cada grupo. Sendo assim, tentando ao máximo ter uma amostra que represente a população. Poderíamos fazer um estudo que prevê se uma árvore será desmatada ou não, baseado em suas características, tempo de vida e localidade em que se encontra.

b. Crianças abaixo de 5 anos e que tiveram sarampo:
Como a população em questão já é bem ninchada (específica), a amostragem aleatória simples é o mais indicado pela sua facilidade de aplicação. A variável a ser estudada poderia ser as sequelas que cada criança teve e a gravidade que a doença causou, para isso, iriamos ter que saber quanto tempo ele esteve no hospital, quanto tempo esteve com a doença, e características físicas e mentais de cada criança

c. Operários em indústrias têxteis: Amostragem estratificada é o melhor tipo de forma de obter uma amostra para esta população, pois separamos a população em grupos e subgrupos e após isso, podemos usar amostragem aleatória simples para cada grupo especificado. Através de entrevistas, poderíamos estudar a satisfação dos funcionários em relação ao trabalho e concluir cada avaliação dos empregados entre três classes (bom, médio ou ruim). Para obter essas informações, os questionários teriam informações pessoais de cada operário (idade, raça, gênero) e informações relacionadas ao trabalho como quanto tempo trabalha nessa indústria, se tem parentes trabalhando no mesmo setor e etc.



# Amostragem Simples

Simplesmente selecionar alguns valores do banco de dados de maneira aleatória. Nesse caso irei selecionar 40% dos meus registros.

In [4]:
amostra = dados.sample(frac=0.40)
amostra.head()

ID  ... Sensor_Proximity(1-highest/10-lowest)
7985  7986  ...                                     7
151    152  ...                                    10
8129  8130  ...                                     5
8248  8249  ...                                     2
8908  8909  ...                                     7

[5 rows x 62 columns]

In [5]:
amostra.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6000 entries, 7985 to 14730
Data columns (total 62 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   ID                                     6000 non-null   int64  
 1   Name                                   5998 non-null   object 
 2   Department                             6000 non-null   object 
 3   GEO                                    6000 non-null   object 
 4   Role                                   6000 non-null   object 
 5   Rising_Star                            7 non-null      float64
 6   Will_Relocate                          6000 non-null   int64  
 7   Critical                               193 non-null    float64
 8   Trending Perf                          6000 non-null   int64  
 9   Talent_Level                           6000 non-null   int64  
 10  Validated_Talent_Level                 6000 non-null   int64  
 11  

Obtendo a amostra, temos 6000 registros, selecionando 40% da população de maneira aleatória. 

# Amostragem Sistemática

Os elementos da população serão colocados em uma lista e cada enésimo elemento da lista é escolhido (sistematicamente) e irá compor a amostra.

In [6]:
seed = np.random.choice(10,1) #selecionando número aleatório de 1 a 10
seed

array([8])

In [7]:
indices = np.arange(0, len(dados), seed)
indices

array([    0,     8,    16, ..., 14976, 14984, 14992])

No primeiro parâmetro do código acima eu seleciono em qual número o array deve se iniciar, no segundo em que número ele deve parar e por último, o intervalo com que os números devem ser escolhidos.


**Obs1**: o método 'np.arange' retorna um array (lista) de valores.

**Obs2**: len(dados) retorna a quantidade total de registros do DataFrame **dados**.

In [8]:
amostra = dados.loc[indices]

No código acima eu localizei, dentro dos meus dados, somente os indíces que correspondem ao meu array 'indices', explicado anteriormente.

**OBS**: Índice de um DataFrame é o primeiro valor do registro da esquerda para direita, que não tem nenhuma label em cima.

In [9]:
amostra.head()

ID  ... Sensor_Proximity(1-highest/10-lowest)
0    1  ...                                     9
8    9  ...                                     7
16  17  ...                                     9
24  25  ...                                     7
32  33  ...                                     9

[5 rows x 62 columns]

In [10]:
amostra.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1875 entries, 0 to 14992
Data columns (total 62 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   ID                                     1875 non-null   int64  
 1   Name                                   1875 non-null   object 
 2   Department                             1875 non-null   object 
 3   GEO                                    1875 non-null   object 
 4   Role                                   1875 non-null   object 
 5   Rising_Star                            2 non-null      float64
 6   Will_Relocate                          1875 non-null   int64  
 7   Critical                               59 non-null     float64
 8   Trending Perf                          1875 non-null   int64  
 9   Talent_Level                           1875 non-null   int64  
 10  Validated_Talent_Level                 1875 non-null   int64  
 11  Per

Como podemos observar, a nossa amostra gerada varia a quantidade de registros de acordo com o número gerado aleatoriamente, selecionando assim, de maneira sistemática.

# Exemplo 2.1

In [17]:
#criando nova amostra com somente 3 registros.
nova_amostra = amostra.sample(n=3)
nova_amostra

ID  ... Sensor_Proximity(1-highest/10-lowest)
14552  14553  ...                                     3
13192  13193  ...                                     8
5088    5089  ...                                     6

[3 rows x 62 columns]

Para os dados descritos na nova amostra, os seguintes parâmetros populacionais podem ser definidos:

i. Departamento de cada funcionário

In [47]:
#Todos os departamentos dos registros obtidos na amostra
nova_amostra['Department'].unique()

array(['Finance', 'IT', 'Operations'], dtype=object)

In [48]:
print('D = {} = Y'.format(nova_amostra['Department'].unique()))

D = ['Finance' 'IT' 'Operations'] = Y


ii. Para o vetor (FiTi)'

In [32]:
# Obtendo os valores da última pontuação da avaliação desse funcionário
nova_amostra.last_evaluation

14552    0.99
13192    0.94
5088     0.61
Name: last_evaluation, dtype: float64

In [40]:
# Obtendo os valores dos números de projetos que o funcionário trabalhou
nova_amostra.number_project

14552    3
13192    7
5088     5
Name: number_project, dtype: int64

B = <br/> 
(0.99  0.94  0.61 <br/> 
3.00  7.00  5.00)<br/> 


Com relação à funções paramétricas populacionais, tem-se:


1. Tempo gasto em média na empresa.

In [39]:
# O método mean() devolve a média aritmética dos números dados a ele. 
nova_amostra['time_spend_company'].mean()

3.3333333333333335

2. Média das variáveis 'last_evaluation' e 'number_project'.

In [44]:
print('média da última avaliação: {:.2f}'.format(nova_amostra['last_evaluation'].mean()))
print('média do número de projetos feitos pelos funcionários {}'.format(nova_amostra['number_project'].mean()))

média da última avaliação: 0.85
média do número de projetos feitos pelos funcionários 5.0


3. Média de vezes que o funcionário visita o LinkedIn por nível de salário.

In [52]:
nova_amostra.groupby(by='salary').mean()['LinkedIn_Hits']

salary
low       10.0
medium    26.5
Name: LinkedIn_Hits, dtype: float64

a. Total populacional


In [62]:
nova_amostra['time_spend_company'].sum()

10

b. Média populacional

In [70]:
nova_amostra['time_spend_company'].mean()

3.3333333333333335

c. Variância populacional

In [72]:
nova_amostra['time_spend_company'].var()

2.3333333333333335

d. Covariância populacional

In [86]:
import numpy as np

X = np.stack((nova_amostra['time_spend_company'], nova_amostra['number_project']), axis=0)
np.cov(X)

array([[2.33333333, 3.        ],
       [3.        , 4.        ]])

e. Correlação populacional

In [94]:
#correlação
corr = nova_amostra.corr()

#valores absolutos
corr_target = abs(corr['time_spend_company'])

#Selecionando as features com maior correlação com a variável de interesse
relevant_features = corr_target[corr_target>0.8]
relevant_features

EMP_Engagement_1                         0.944911
EMP_Engagement_2                         0.944911
EMP_Engagement_3                         0.944911
EMP_Engagement_4                         0.944911
EMP_Engagement_5                         0.944911
number_project                           0.981981
average_montly_hours                     0.974248
time_spend_company                       1.000000
left_Company                             0.944911
Emp_Competitive_2                        0.944911
Emp_Competitive_4                        0.981981
Emp_Collaborative_1                      0.981981
Emp_Collaborative_3                      0.891042
Emp_Collaborative_5                      0.981981
Sensor_StepCount                         0.907171
Sensor_Heartbeat(Average/Min)            0.907841
Sensor_Proximity(1-highest/10-lowest)    0.953821
Name: time_spend_company, dtype: float64

f. Razão populacional

In [102]:
x = nova_amostra['time_spend_company'].sum()
y = nova_amostra['number_project'].sum()
r_populacional = x/y
r_populacional

0.6666666666666666

g. Razão média populacional

In [103]:
r_populacional/3

0.2222222222222222